File: GuideObjC/GuideObjC/Managers/OffensivePosts.m
//
//  OffensivePosts.m
//  GuideObjC
//
//  Created by LuzanovRoman on 30.12.16.
//  Copyright © 2016 eigengraph. All rights reserved.
//

#import "OffensivePosts.h"

@interface OffensivePosts ()
@property (retain, nonatomic) NSMutableArray <NSString *> *notOffensivePostsIds;
@property (retain, nonatomic) NSMutableArray <NSString *> *offensivePostsIds;
@property (retain, nonatomic) NSMutableArray <NSString *> *checkingIds;
@property (retain, nonatomic) NSMutableArray <NSString *> *addingIds;
@property (retain, nonatomic) NSString *currentUserId;
@property (retain, nonatomic) NSString *edge;
@end

@implementation OffensivePosts

+ (OffensivePosts *)shared {
    static OffensivePosts * sharedInstance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedInstance = [[OffensivePosts alloc] init];
    });
    return sharedInstance;
}

- (id)init {
    self = [super init];
    
    if (self) {
        _edge = @"offended";
        _addingIds = [NSMutableArray array];
        _checkingIds = [NSMutableArray array];
        _offensivePostsIds = [NSMutableArray array];
        _notOffensivePostsIds = [NSMutableArray array];
    }
    return self;
}

- (void)startSession {
    [[Graph shared] userObjectWithCompletion:^(NSObject * object, NSError * error) {
        if ([object isMemberOfClass:[EGFUser class]]) {
            _currentUserId = ((EGFUser *)object).id;
        }
    }];
}

- (void)stopSession {
    [_notOffensivePostsIds removeAllObjects];
    [_offensivePostsIds removeAllObjects];
    [_checkingIds removeAllObjects];
    [_addingIds removeAllObjects];
    _currentUserId = nil;
}

- (OffensiveState)offensiveStateForPost:(EGFPost*)post {
    if (!post.id || !_currentUserId) {
        return osUnknown;
    }
    if ([_checkingIds containsObject:post.id]) {
        return osUnknown;
    }
    if ([_addingIds containsObject:post.id] || [_offensivePostsIds containsObject:post.id]) {
        return osOffensive;
    }
    if ([_notOffensivePostsIds containsObject:post.id]) {
        return osNotOffensive;
    }
    [_checkingIds addObject:post.id];
    [[Graph shared] doesObjectWithId:_currentUserId existForSource:post.id onEdge:_edge completion:^(BOOL exists, NSError * error) {
        [_checkingIds removeObject:post.id];
        
        if (!error) {
            if (exists) {
                [_offensivePostsIds addObject:post.id];
            }
            else {
                [_notOffensivePostsIds addObject:post.id];
            }
        }
    }];
    return osUnknown;
}

- (void)markAsOffensive:(EGFPost *)post completion:(void (^)())completion {
    if (!post.id || !_currentUserId) {
        completion();
        return;
    }
    [_addingIds addObject:post.id];
    [[Graph shared] addObjectWithId:_currentUserId forSource:post.id toEdge:_edge completion:^(id object, NSError * error) {
        [_addingIds removeObject:post.id];
        
        if (!error) {
            [_offensivePostsIds addObject:post.id];
            [_notOffensivePostsIds removeObject:post.id];
        }
        completion();
    }];
}
@end

File: GuideSwift/GuideSwift/Managers/OffensivePosts.swift
//
//  OffensivePosts.swift
//  GuideSwift
//
//  Created by LuzanovRoman on 29.12.16.
//  Copyright © 2016 eigengraph. All rights reserved.
//

import Foundation

enum OffensiveState {
    case isNotOffensive
    case isOffensive
    case isUnknown
}

class OffensivePosts {
    static let shared = OffensivePosts()

    fileprivate var notOffensivePostsIds = [String]()
    fileprivate var offensivePostsIds = [String]()
    fileprivate var checkingIds = [String]()
    fileprivate var addingIds = [String]()
    fileprivate var currentUserId: String?
    fileprivate let edge = "offended"

    func startSession() {
        Graph.userObject { (object, _) in
            self.currentUserId = (object as? EGFUser)?.id
        }
    }

    func stopSession() {
        notOffensivePostsIds.removeAll()
        offensivePostsIds.removeAll()
        checkingIds.removeAll()
        addingIds.removeAll()
        currentUserId = nil
    }

    func offensiveState(forPost post: EGFPost) -> OffensiveState {
        guard let postId = post.id, let userId = currentUserId else { return .isUnknown }

        if checkingIds.contains(postId) {
            return .isUnknown
        }
        if addingIds.contains(postId) || offensivePostsIds.contains(postId) {
            return .isOffensive
        }
        if notOffensivePostsIds.contains(postId) {
            return .isNotOffensive
        }
        checkingIds.append(postId)
        Graph.doesObject(withId: userId, existForSource: postId, onEdge: edge) { (exists, error) in
            self.checkingIds.remove(postId)

            if let _ = error { return }

            if exists {
                self.offensivePostsIds.append(postId)
            } else {
                self.notOffensivePostsIds.append(postId)
            }
        }
        return .isUnknown
    }

    func markAsOffensive(post: EGFPost, completion: @escaping () -> Void) {
        guard let postId = post.id, let userId = currentUserId else {
            completion()
            return
        }
        addingIds.append(postId)
        Graph.addObject(withId: userId, forSource: postId, toEdge: edge) { (_, error) in
            self.addingIds.remove(postId)

            if error == nil {
                self.offensivePostsIds.append(postId)
                self.notOffensivePostsIds.remove(postId)
            }
            completion()
        }
    }
}
